Sources/VLCOneDriveController.m

```diff
@@ -160,68 +158,189 @@ - (void)authFailed:(NSError *)error userState:(id)userState
+}
 
-- (void)requestDirectoryListingAtPath:(NSString *)path
+- (void)loadODItems
 {
-    [self loadCurrentFolder];
+    [self loadODItemsWithCompletionHandler:nil];
 }
 
-- (void)loadTopLevelFolder
+- (void)loadThumbnails:(NSArray<ODItem *> *)items
 {
-    _rootFolder = [[VLCOneDriveObject alloc] init];
-    _rootFolder.objectId = @"me/skydrive";
-    _rootFolder.name = @"OneDrive";
-    _rootFolder.type = @"folder";
-    _rootFolder.liveClient = _liveClient;
-    _rootFolder.delegate = self;
-
-    _currentFolder = _rootFolder;
-    [_rootFolder loadFolderContent];
+    for (ODItem *item in items) {
+        if ([item thumbnails:0]) {
+            [[[[[_oneDriveClient.drive items:item.id] thumbnails:@"0"] small] contentRequest]
+             downloadWithCompletion:^(NSURL *location, NSURLResponse *response, NSError *error) {
+                 if (!error) {
+                 }
+             }];
+        }
+    }
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (self.delegate) {
+            [self.delegate performSelector:@selector(mediaListUpdated)];
+        }
+    });
+}
+
+#pragma - subtitle
+
+- (NSString *)configureSubtitleWithFileName:(NSString *)fileName folderItems:(NSArray *)folderItems
+{
+    NSString *subtitleURL = nil;
+    NSString *subtitlePath = [self _searchSubtitle:fileName folderItems:folderItems];
+
+    if (subtitlePath)
+        subtitleURL = [self _getFileSubtitleFromServer:[NSURL URLWithString:subtitlePath]];
+
+    return subtitleURL;
+}
+
+- (NSString *)_searchSubtitle:(NSString *)fileName folderItems:(NSArray *)folderItems
+{
+    NSString *urlTemp = [[fileName lastPathComponent] stringByDeletingPathExtension];
+    NSString *itemPath = nil;
+
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"name contains[c] %@", urlTemp];
+    NSArray *results = [folderItems filteredArrayUsingPredicate:predicate];
+
+    for (ODItem *item in results) {
+        if ([item.name isSupportedSubtitleFormat]) {
+            itemPath = item.dictionaryFromItem[@"@content.downloadUrl"];
+        }
+    }
+    return itemPath;
 }
 
-- (void)loadCurrentFolder
+- (NSString *)_getFileSubtitleFromServer:(NSURL *)subtitleURL
 {
-    if (_currentFolder == nil)
-        [self loadTopLevelFolder];
-    else {
-        _currentFolder.delegate = self;
-        [_currentFolder loadFolderContent];
+    NSString *FileSubtitlePath = nil;
+    NSData *receivedSub = [NSData dataWithContentsOfURL:subtitleURL]; // TODO: fix synchronous load
+
+    if (receivedSub.length < [[UIDevice currentDevice] VLCFreeDiskSpace].longLongValue) {
+        NSArray *searchPaths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
+        NSString *directoryPath = searchPaths[0];
+        FileSubtitlePath = [directoryPath stringByAppendingPathComponent:[subtitleURL lastPathComponent]];
+
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+        if (![fileManager fileExistsAtPath:FileSubtitlePath]) {
```

Comment:
Note

Attempting to predicate behavior based on the current state of the file system or a particular file on the file system is not recommended. Doing so can cause odd behavior or race conditions. It’s far better to attempt an operation (such as loading a file or creating a directory), check for errors, and handle those errors gracefully than it is to try to figure out ahead of time whether the operation will succeed.

Reply:
Code that didn't change from VLCODObject but I can change it.

Reply:
up to you :)